db/migrate/20150716133712_add_unique_index_on_journals.rb

```diff
@@ -0,0 +1,99 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddUniqueIndexOnJournals < ActiveRecord::Migration
+  def up
+    cleanup_duplicate_journals
+    add_index :journals, [:journable_type, :journable_id, :version], unique: true
+  end
+
+  def down
+    remove_index :journals, [:journable_type, :journable_id, :version]
+  end
+
+  private
+
+  def cleanup_duplicate_journals
+    duplicate_pairs = find_duplicate_journal_ids
+
+    if duplicate_pairs.any?
+      say "Found #{duplicate_pairs.count} journals with at least one duplicate!"
+      say_with_time 'Safely removing duplicates...' do
+        duplicate_pairs.each do |current_id, duplicate_id|
+          say "Comparing journals ##{current_id} & ##{duplicate_id} for equality", subitem: true
+
+          current = Journal.find(current_id)
+          duplicate = Journal.find(duplicate_id)
+
+          if journals_equivalent?(current, duplicate)
+            say "Deleting journal ##{current.id}...", subitem: true
+            current.destroy
```

Comment:
This happens when you destroy a Journal:

![image](https://cloud.githubusercontent.com/assets/617519/8729960/7270059e-2bee-11e5-8b45-fa246e600bb3.png)

The actual data (work_package_journals) is not removed. The same happens BTW if you delete a WorkPackage (or other content I suppose).

This means that we probably have quite an amount of unconnected journal data objects in our DB. This should pose no immediate problem but it will clutter the Database.

Not a problem of this PR, though. I will open a ticket on OP.org.